lib/livebook_web/live/session_live/secrets_component.ex

```diff
@@ -322,23 +368,33 @@ defmodule LivebookWeb.SessionLive.SecretsComponent do
     end
   end
 
-  defp grant_access(secret_name, socket) do
-    secret_value = socket.assigns.livebook_secrets[secret_name]
-    secret = %{name: secret_name, value: secret_value}
-    set_secret(socket.assigns.session.pid, secret, "session")
-  end
+  defp grant_access(secrets, secret_name, origin, socket) do
+    secret = Enum.find(secrets, &(&1.name == secret_name and &1.origin == origin))
 
-  defp livebook_only_secrets(secrets, livebook_secrets) do
-    Enum.reject(livebook_secrets, &(&1 in secrets)) |> Enum.sort()
+    if secret,
+      do: set_secret(socket, secret),
+      else: :ok
   end
 
-  defp must_grant_access(%{assigns: %{prefill_secret_name: prefill_secret_name}} = socket) do
-    if not Map.has_key?(socket.assigns.secrets, prefill_secret_name) and
-         Map.has_key?(socket.assigns.livebook_secrets, prefill_secret_name) do
-      prefill_secret_name
+  defp must_grant_access(%{assigns: %{prefill_secret_name: secret_name}} = socket) do
+    if not session?(socket, secret_name) and
+         (livebook?(socket, secret_name) or hub?(socket, secret_name)) do
+      secret_name
     end
   end
 
+  defp session?(socket, secret_name) do
+    Enum.any?(socket.assigns.secrets, &(elem(&1, 0) == secret_name))
+  end
+
+  defp livebook?(socket, secret_name) do
+    Enum.any?(socket.assigns.livebook_secrets, &(&1.name == secret_name and &1.origin == :app))
+  end
+
+  defp hub?(socket, secret_name) do
+    Enum.any?(socket.assigns.hub_secrets, &(&1.name == secret_name and is_binary(&1.origin)))
```

Comment:
```suggestion
    Enum.any?(socket.assigns.hub_secrets, &match?(%{name: ^secret_name, origin: {:hub, _}}, &1))
```

although do we need to check origin if we are looking specifically at `hub_secrets`?

Reply:
Yes, because these three functions (`session?/1`, `livebook?/1`, `hub?/1`) checks the given secret in every possibly source

Reply:
Oh, sorry. We don't really need to check in this case, ur right. Good call